src/main/java/games/chess/ChessGameState.java

```diff
@@ -0,0 +1,414 @@
+        copy.blackPieces.addAll(blackPieces.stream().map(ChessPiece::copy).toList());
+        copy.gameStateCounts = new HashMap<>(gameStateCounts);
+        copy.halfMoveClock = halfMoveClock;
+        copy.board = (ChessBoard) board.copy();
+        return copy;
+    }
+
+
+    @Override
+    protected double _getHeuristicScore(int playerId) {
+        if (isNotTerminal()) {
+            // Simple value for each piece on the board, weighted by its type. Find the difference between the two players.
+            double playerScore = 0.0; 
+            double opponentScore = 0.0;
+            for (ChessPiece piece : getPlayerPieces(1 - playerId)) {   
+                ChessPiece.ChessPieceType type = piece.getChessPieceType();
+                switch (type) {
+                    case PAWN:
+                        opponentScore += 1;
+                        break;
+                    case KNIGHT:
+                    case BISHOP:
+                        opponentScore += 3;
+                        break;
+                    case ROOK:
+                        opponentScore += 5;
+                        break;
+                    case QUEEN:
+                        opponentScore += 9;
+                        break;
+                    case KING:
+                        opponentScore += 0;
+                        break;
+                }
+            }
+            for (ChessPiece piece : getPlayerPieces(playerId)) {
+                ChessPiece.ChessPieceType type = piece.getChessPieceType();
+                switch (type) {
+                    case PAWN:
+                        playerScore += 1;
+                        break;
+                    case KNIGHT:
+                    case BISHOP:
+                        playerScore += 3;
+                        break;
+                    case ROOK:
+                        playerScore += 5;
+                        break;
+                    case QUEEN:
+                        playerScore += 9;
+                        break;
+                    case KING:
+                        playerScore += 0;
+                        break;
+                }
+            }
+            //Reward check
+            if (isInCheck(1-playerId)) {
+                playerScore += 1;
+            }
+            //Penalize oppenent king movement choices
+            int opponentKingActions = computeAvailableActionsKing(getKingPosition(1-playerId)[0], getKingPosition(1-playerId)[1], 1-playerId).size();
+            playerScore -= opponentKingActions * 0.2; // TODO: Tune this value to be more or less punishing
+
+            return (playerScore - opponentScore)/100; // TODO: Normalize this value to be between 0 and 1
+        } else {
+            // The game finished, we can instead return the actual result of the game for the given player.
+            return getPlayerResults()[playerId].value;
+        }
+    }
+
+    @Override
+    public double getGameScore(int playerId) {
+        return 0;
```

Comment:
This should check for a terminal game state, in which case the score is 0, 0.5, 1 for loss/draw/win?